annot-tsv.c

```diff
@@ -315,47 +316,90 @@ void free_payload(void *payload)
     cols_destroy(cols);
 }
 
-// Parse header if present (first line has a leading #) or create a dummy header with
-// numeric column names. If dummy is set, read first data line (without a leading #)
-// and create a dummy header.
-void parse_header(dat_t *dat, char *fname, int dummy)
+// Parse header if present, the parameter irow indicates the header row line number:
+//      0   .. ignore headers, create numeric fields names, 1-based indices
+//      N>0 .. N-th line, all previous lines are discarded
+//      N<0 .. N-th line from the end of the comment block (comment lines are prefixed with #),
+//             all preceding lines are discarded.
+// When autodetect is set, the argument nth_row is ignored.
+// Note this makes no attempt to preserve comment lines on output
+void parse_header(dat_t *dat, char *fname, int nth_row, int autodetect)
 {
     dat->fp = hts_open(fname,"r");
     if ( !dat->fp ) error("Failed to open: %s\n", fname);
 
+    // buffer comment lines when N<0
+    int nbuf = 0;
+    char **buf = NULL;
+    if ( nth_row < 0 )
+    {
+        buf = calloc(-nth_row,sizeof(*buf));
+        if ( !buf ) error("Out of memory, failed to allocate %zu bytes\n",(-nth_row)*sizeof(*buf));
+    }
+
+    int irow = 0;
     cols_t *cols = NULL;
     while ( hts_getline(dat->fp, KS_SEP_LINE, &dat->line) > 0 )
     {
-        if ( dat->line.s[0]=='#' )
+        if ( autodetect )
+        {
+            // if the first line is comment line, use it as a header. Otherwise go
+            // with numeric indices
+            nth_row = dat->line.s[0]=='#' ? 1 : 0;
+            break;
+        }
+        if ( nth_row==0 )
         {
-            // this is a header or comment line
-            if ( dummy ) continue;
-            cols = cols_split(dat->line.s, NULL, '\t');
+            // N=0 .. comment lines to be ignored, read until we get to the first data line
+            if ( dat->line.s[0]=='#' ) continue;
             break;
         }
+        if ( nth_row>0 )
+        {
+            // N>1 .. regardless of this being a comment or data line, read until Nth line
+            if ( ++irow < nth_row ) continue;
+            break;
+        }
+        // N<0 .. keep abs(N) comment lines in a sliding buffer
+        if ( dat->line.s[0]!='#' ) break;   // data line
+        if ( nbuf == -nth_row )
```

Comment:
With a negative in the nth_row the first line of the actual data is missing in the output and memory is leaked from the strdup.